#161342839 Enable Social Login via Google, Facebook and twitter

ucheg6 opened the pull request:
#### What does this PR do?
- Enable Social Login via Google, Facebook and twitter
#### Description of Task to be completed?
- setup the Authors haven app as a project on facebook and google developer websites
- configure facebook and google strategy using passportJs
- set up auth/facebook and auth/google routes to authenticate social media users
#### How should this be manually tested?
- git clone ` https://github.com/andela/haven-ah-backend.git `
- run npm install to install necessary npm packages
- create dot-env file and setup necessary variables for client secret and client key for facebook and google
-  run ` createdb haven-ah `to create the database
-  run ` sequelize:migrate ` to migrate model
-  navigate to the app's home route and also the social media login routes specified in socialLogin.js file
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#161342839](https://www.pivotaltracker.com/story/show/161342839)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

victor-nweze-deel commented on app.js:
```diff
@@ -21,10 +23,16 @@ app.use(logger('dev'));
 app.use(cleanStrings);
-
+app.use(session({
+  secret: 's3cr3t',
```
This should be in the environment variables

ucheg6 replied:
This is true. I'll implement this ASAP

victor-nweze-deel commented on app.js:
```diff
@@ -21,10 +23,16 @@ app.use(logger('dev'));
 router(app);
 
-app.get('*', (request, response) => {
+app.get('/', (request, response) => {
```
Why this?

ucheg6 replied:
I figured that the * picks invalid routes and renders the welcome message. I have created another function to handle both cases.

victor-nweze-deel commented on server/routes/index.js:
```diff
@@ -1,7 +1,9 @@
 export default (app) => {
   app.use('/api/v1', userRouter);
   app.use('/api/v1/articles', articleRouter);
+  app.use('/api/v1', socialLoginRouter);
```
Priscilla has another implementation of this. Please revisit and refactor accordingly.

ucheg6 replied:
I'll fix this when I'm resolving merge conflicts.

seyi-adeleke commented on app.js:
```diff
@@ -1,11 +1,13 @@
 import express from 'express';
 import bodyParser from 'body-parser';
 import logger from 'morgan';
+import session from 'express-session';
```
Yo, we're not using sessions in this application.

ucheg6 replied:
Twitter uses Oauth1.0 and it enforces express-session middleware. It's there because of twitter other implementations use oauth2.0

seyi-adeleke commented on server/repository/userRepository.js:
```diff
@@ -157,6 +158,34 @@ class UserRepository {
+      }
+    })
+      .spread((user, created) => {
+        console.log('created=======>', created);
```
remove this log

ucheg6 replied:
OK. I will

seyi-adeleke commented on server/repository/userRepository.js:
```diff
@@ -157,6 +158,34 @@ class UserRepository {
+  /**
+ * finds a user by email, if user exists return the user, if not create one
+ * @param {string} email Email to search by
+ * @param {string} familyName familyName to insert to lastName column in database
```
why cant familyName be lastName?

ucheg6 replied:
Ok. I will change this

seyi-adeleke commented on server/repository/userRepository.js:
```diff
@@ -157,6 +158,34 @@ class UserRepository {
+ * finds a user by email, if user exists return the user, if not create one
+ * @param {string} email Email to search by
+ * @param {string} familyName familyName to insert to lastName column in database
+ * @param {string} givenName givenName to insert to firstName column in database
```
Same comment as above

seyi-adeleke commented on server/routes/index.js:
```diff
@@ -2,12 +2,14 @@ import { Router } from 'express';
 
 import userRouter from './users';
 import articleRouter from './article';
+import socialLoginRouter from './socialLogin';
```
No need to create a seperate router for this.  We have some authentication routes within userRouter so you can add it there

ucheg6 replied:
Ok. I will make the change

seyi-adeleke commented on server/utilities/passportAuthentication.js:
```diff
@@ -0,0 +1,10 @@
+import passport from 'passport';
+
+export default (socialMediaName) => {
```
jsdoc and tests please

ucheg6 replied:
Ok.. I'm working on this

seyi-adeleke replied:
has this been addressed?

ucheg6 replied:
Yes it has

seyi-adeleke commented on app.js:
```diff
@@ -23,12 +25,24 @@ app.use(bodyParser.json());
+  saveUninitialized: true
+}));
+app.use(passport.initialize());
+app.use(passport.session());
```
Remove line 29 and 35

seyi-adeleke commented on app.js:
```diff
@@ -23,12 +25,24 @@ app.use(bodyParser.json());
 app.use('/api/v1', router);
 
-app.get('*', (request, response) => {
+app.get('/', (request, response) => {
```
What happens if i make a `post` request to the '/' route?

ucheg6 replied:
The / route picks up the home route while the * route picks up every route that does not exist.

seyi-adeleke commented on package.json:
```diff
@@ -9,7 +9,7 @@
     "build": "npm run clean && mkdir dist/ && npm run babel-server && npm run babel-app",
     "clean": "rm -rf dist/",
-    "coveralls": "nyc --reporter=lcov --reporter=text --require babel-register  mocha server/tests/**/*.test.js --exit",
+    "coveralls": "nyc --reporter=lcov --reporter=text --require babel-register  mocha --timeout 15000 server/tests/**/*.test.js --exit",
```
why do we need this here?

ucheg6 replied:
After I resolved merge conflicts I discovered that coveralls wasn't reporting so I met Victor and he suggested I add the timeout. That got coveralls to show report.

seyi-adeleke commented on server/migrations/20181023192004-create-user.js:
```diff
@@ -9,7 +9,7 @@ export default {
     },
     username: {
       type: Sequelize.STRING,
```
whats happening here?

ucheg6 replied:
I had already effected this changes directly on the migration file before you gave us the feedback to always create a different migration file. I changed the null constraints because I'm not getting some user details from social media so I needed to change the constraints to save user data in the database

seyi-adeleke commented on server/services/facebookStrategy.js:
```diff
@@ -0,0 +1,29 @@
+
+dotenv.config();
+
+passport.use(new FacebookStrategy({
```
Can we have the social configs in the config directory? Create them there then import into this file

ucheg6 replied:
Ok will do this

seyi-adeleke commented on server/services/twitterStrategy.js:
```diff
@@ -0,0 +1,47 @@
+const { User } = Model;
+
+dotenv.config();
+passport.serializeUser((user, done) => {
```
whats this for?

ucheg6 replied:
Yeah, It's for the express-session twitter's oauth1.0 is enforcing. It works with the session to serialise and deserialise user data.

seyi-adeleke commented on server/config/config.js:
```diff
@@ -23,6 +23,16 @@ const config = {
     dialect: 'postgres',
     secret: process.env.JWT_SECRET
   },
+  facebook: {
```
nope. lets leave our environment config as is. You can create a new file for facebook and google. Sorry I wasn't clear

ucheg6 replied:
Ok... I will effect this change now.

seyi-adeleke commented on server/repository/userRepository.js:
```diff
@@ -205,6 +205,33 @@ class UserRepository {
+  /**
+ * finds a user by email, if user exists return the user, if not create one
+ * @param {string} email Email to search by
+ * @param {string} lastName familyName to insert to lastName column in database
```
familyName and givenName still?

ucheg6 replied:
This has been handled as well

seyi-adeleke commented:
Also, no documentation?

seyi-adeleke commented on package.json:
```diff
@@ -28,17 +28,21 @@
     "cloudinary": "^1.11.0",
     "dotenv": "^6.1.0",
     "express": "^4.16.4",
+    "express-session": "^1.15.6",
```
Are we still using this package?

seyi-adeleke commented on server/services/socialCallback.js:
```diff
@@ -0,0 +1,12 @@
+import userRepo from '../repository/userRepository';
+
+const socialCallback = (accessToken, refreshToken, profile, done) => {
```
jsdoc

ucheg6 replied:
This is done

ucheg6 committed:
feat(social Login): Enable Social Login via Google, Facebook and twitter

- setup the Authors haven app as a project on facebook and google developer websites
- configure facebook and google strategy using passportJs
- set up auth/facebook and auth/google routes to authenticate social media users
